### ISSUE TITLE
Missing support for i2d systems (iqpump01)

### ISSUE THREAD

**Issue by aruiz760** [flz/iaqualink-py#36]
Hello, I hope I'm putting this in the correct location.  The Jandy  IQPUMP01 shows up in the Iaqualink app but not on the integration.  Here is a log output:

_Logger: homeassistant.components.iaqualink
Source: components/iaqualink/__init__.py:98
Integration: Jandy iAqualink (documentation, issues)
First occurred: 12:28:20 AM (8 occurrences)
Last logged: 1:28:00 PM

No systems detected or supported_

Would appreciate any help.  Thanks.

**Comment by Joacbe** [comment]
IQPUMP seems a bit special and is not supported out of the box by the module apparently (at least I was taking a look at the Home Assistant integration). 

I don't have the skills to modify anything of the existing but I've been able to identify the various API URLs to be called to perform the basic operations and validate them all on my pupm. 

I'll be more than happy to share if anybody has the skills and is willing to add the IQPUMP01 support.

**Comment by aruiz760** [comment]
Thanks.

**Comment by robertw** [comment]
> IQPUMP seems a bit special and is not supported out of the box by the module apparently (at least I was taking a look at the Home Assistant integration).
> 
> I don't have the skills to modify anything of the existing but I've been able to identify the various API URLs to be called to perform the basic operations and validate them all on my pupm.
> 
> I'll be more than happy to share if anybody has the skills and is willing to add the IQPUMP01 support.

If you have the api calls and willing to share that would be a big help thnak you

**Comment by Joacbe** [comment]
I can share my Postman project where all the details are aggregated.
[ZodiacIQPump.postman_collection.json](https://github.com/flz/iaqualink-py/files/15030878/ZodiacIQPump.postman_collection.json)


### I have 8 variables defined:
![image](https://github.com/flz/iaqualink-py/assets/141075230/6ee23064-f682-4e59-abfb-321bc1c52da8)

Below variables values are coming from your Jandy/iAqualink account:

- JandyUserLogin
- JandyUserPassword
- APIKey = **EOOEMOW4YR6QNB07** (apparently it never changes)
- PumpSerial = the serial number without the dashes signs
![image](https://github.com/flz/iaqualink-py/assets/141075230/f49081bf-f83a-46f4-b260-d0d5401f1d5b)
Note: the serial number may also be retrieved using the query named "Get Devices" as the "serial_number" attribute in the returned json.


### And identified 10 different API calls 
Those are allowing to perform all the most important operations over the pump:
![image](https://github.com/flz/iaqualink-py/assets/141075230/da0ef481-3529-48f7-a2b2-47f5ba431453)

- [ ] The Auth & GetToken query allows connecting and getting the required values for defining the following variables:

- IDToken = IdToken
- AccessToken = AccessToken
- AuthToken = authentication_token
- UserID = id

- [ ] Get Devices allows getting all the iAqualink devices existing in association to your account. In my case, I only have the pump, device of type i2d

- [ ]  0 - Get Device All Data

### Dealing with pump override/forced activation start / stop
- [ ]  1 - Device Write Mode - Set custom time
- [ ]  2 - Device Write - Custom Speed
- [ ]  3 - Device Write - Custom Operations time
- [ ]  4a - Device Stop pump 

### Setting back pump in scheduled operating mode
- [ ]  4b - Device Start pump - Resume Schedule

### Dealing with pump schedules: reading/writing
- [ ]  5 - Device Read Schedules
- [ ]  6 - Device Write Schedules


Last but not least, this is how to interprete the values returned when dealing with the schedules:

- Returned data on a single row: Y070018001500YYYYYYYDefault Schedule
 = Y : is schedule enabled? Y / N
 = 0700 : start time (always 4 digits)
 = 1800 : stop time (always 4 digits)
 = 1500 : RPM during the schedule (always 4 digits, filled in with 0 if less than 1000)
 = YYYYYYY : week days the schedule is enabled. Could be N as well if disabled. Note: starting with Sunday!!
 = Default Schedule: name of the schedule

Notes, schedules are all separated by a | and stored on a single line

**Comment by robertw** [comment]
Many thanks i am not a great coder but will try to make it work with home Assistant.......

**Comment by unc2k3** [comment]
Just wanted to see if anyone can walk me through how to set this up.  I've also got an iqpump01 controller for my pump and I cannot figure out how to get this to show up in the iAquaLink integration in Home Assistant.

**Comment by flz** [comment]
Hey, that's a good start! A few comments:
- Let's put this in a branch/PR to facilitate review.
- This library largely exists to hide the implementation details from home-assistant. The device classes that should be created should map to the type of entities you'd want to have available in HA. There was a proposal for a new pump entity but this hasn't gone anywhere.
  - Since this is only reading information right now, you might want to start by making this a sensor class and exposing all those @properties as attributes.
- Let's make sure to cover the code in unit tests once you've finalized the API.

**Comment by flz** [comment]
Update: I've just bought a used iqpump01 off ebay so I should be able to do some testing once it's installed.

**Comment by unc2k3** [comment]
> Update: I've just bought a used iqpump01 off ebay so I should be able to do some testing once it's installed.

Looking forward to this and any updates.  Would be great to finally get this working!

**Comment by flz** [comment]
I've added the code that @tkhduracell submitted above to a new branch (PR [here](https://github.com/flz/iaqualink-py/pull/85)).

Still needs some cleaning up as mentioned above.

**Comment by CLARENNE-Q** [comment]
Hello,

You can use this projet: https://github.com/CLARENNE-Q/iaqualink_iqpump01

**Comment by dbrownad** [comment]
Thanks for the integration. When I tried to log in after adding it to HACS and Hassio it give me a unable to connect error. I am able to log into the app and have even changed my password but the same error still persists. any ideas would be appreciated.

So far I have:
- changed my password.
- logged onto "https://iaqualink.zodiacpoolsystems.com/signin?lang=en"
- restarted Home Assistant.

**Comment by CLARENNE-Q** [comment]
Hi! 👋

Thanks for trying out the integration and for the detailed steps you've already taken — that's super helpful!

I just released a new version (1.0.3) of the integration which includes:

-     Improved error logging
-     More detailed debug information
-     Better handling of login sessions

Can you please update to the latest version and try again?

If the problem persists, feel free to open a new issue in the project and enable debug logs so I can help you further. You can enable logging and share the relevant output (just be sure to remove your credentials/tokens before sharing anything).

**Comment by dbrownad** [comment]
I installed the 1.0.3 and it authenticated perfectly, Thanks so much, for the work! I have a salt cell that uses a smart 2 pole switch and have needed this for automation.

**Comment by dbrownad** [comment]
I will use this integration to see if the pump is running in order to turn the salt cell on. One thing I noticed, if i try to turn on the pump with the power in HASSIO, it immediatly turns off. maybe a slight adjustment will correct the issue. Thanks again!

**Comment by CLARENNE-Q** [comment]
Yes, I’ve noticed the same behavior — the pump state isn’t updated immediately after changing it via Home Assistant. After checking, it looks like even the iAquaLink mobile app doesn’t refresh the pump status in real time. It seems like the API itself has a built-in delay of about a minute before reflecting the new state. So it’s not an issue with Home Assistant but rather how the backend operates

**Comment by robertw** [comment]
I agree it's either a timing issue or not recognizing a command coming back or something of that nature or some delay I'll try to pull  some logs thanks

**Comment by CLARENNE-Q** [comment]
You can do the job ;-)

**Comment by flz** [comment]
> You can do the job ;-)

Would you be open to maintaining the code in this repo if somebody did the work?

**Comment by TheNetworker69** [comment]
Hello , I installed the Clarenne module on Home Assistant, it sees my iqpump system, it is able to return all the information but it is impossible to change the pump speed via Home Assistant, have you ever had this problem?

<img width="343" height="90" alt="Image" src="https://github.com/user-attachments/assets/ed57e4ed-4ae4-4ac6-be7d-c65be3f57334" />

**Comment by CLARENNE-Q** [comment]
> Hello , I installed the Clarenne module on Home Assistant, it sees my iqpump system, it is able to return all the information but it is impossible to change the pump speed via Home Assistant, have you ever had this problem?
> <img alt="Image" width="343" height="90" src="https://private-user-images.githubusercontent.com/154514695/584116616-ed57e4ed-4ae4-4ac6-be7d-c65be3f57334.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Nzg2NTMxMTMsIm5iZiI6MTc3ODY1MjgxMywicGF0aCI6Ii8xNTQ1MTQ2OTUvNTg0MTE2NjE2LWVkNTdlNGVkLTRhZTQtNGFjNi1iZTdkLWM2NWJlM2Y1NzMzNC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwNTEzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDUxM1QwNjEzMzNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1mMjIxOTJlY2ZlZjFkYTliZTFlZDM3ZDJjZmMwMjdjZTAwZTk2NGRiNjIwNWUyOTYzNGEyNjhkNGIwZGJlN2RlJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZyZXNwb25zZS1jb250ZW50LXR5cGU9aW1hZ2UlMkZwbmcifQ.QKxNF1EcgvALR8vY5yLEIZ2R6UA0SnwRw-p4QW3lJho">

It's fixed in new version ;-)

**Comment by TheNetworker69** [comment]
La nouvelle version fonctionne parfaitement ! Merci

**Comment by flz** [comment]
I'm going to start working on iqpump01 support in the next week or so (trying to wrap up some other things first, including onetouch support).

I invite people using @CLARENNE-Q's custom integration to direct bug reports to his repository and keep this thread for future iqpump01 support in the iaqualink-py library.

I'll post again when something is ready for testing!

**Comment by CLARENNE-Q** [comment]
> I'm going to start working on iqpump01 support in the next week or so (trying to wrap up some other things first, including onetouch support).
> 
> I invite people using @CLARENNE-Q's custom integration to direct bug reports to his repository and keep this thread for future iqpump01 support in the iaqualink-py library.
> 
> I'll post again when something is ready for testing!



You can Check docs/FIELD_NOTES.md for moré info !

**Comment by flz** [comment]
I've now landed support in master. If you're interested in future updates for i2d systems, please subscribe to this [discussion](https://github.com/flz/iaqualink-py/discussions/204) as I will be using it to ask for testing on occasion.

**Comment by thestealthsmarthome** [comment]
For anyone landing here who just wants to *see* their iQPUMP01 pump in Home Assistant (RPM, power, energy) rather than control it from the app, I went a different route that sidesteps the i2d problem entirely: a passive, read-only RS485 tap.

An ESP32 (M5 ATOM + RS485 base) listens on the iQPUMP01-to-pump bus and parses the demand frames into HA sensors via ESPHome. It never transmits, so the iAqualink app keeps full control and there's no fighting the controller for the single-master bus.

What it exposes: RPM (exact, from the master demand frame), running state, and energy. Watts isn't on the bus at all (the app models it from RPM), so I model it the same way and flag it as an estimate.

Repo with wiring, the decoded frame format, and the full ESPHome config:
https://github.com/thestealthsmarthome/esphome-jandy-pool-pump-monitor

This won't help if you need app-style control from HA (that still needs i2d support in the library), but if monitoring is all you're after, it works today on a FloPro + iQPUMP01.
